In [70]:
import pandas as pd
import numpy as np

from tqdm import tqdm_notebook

from sklearn.feature_extraction.text import TfidfTransformer, CountVectorizer, TfidfVectorizer
from sklearn.neighbors import NearestNeighbors

%matplotlib inline

In [71]:
links = pd.read_csv('links.csv')
movies = pd.read_csv('movies.csv')
moviess = pd.read_csv('movies.csv')
ratings = pd.read_csv('ratings.csv')
tags = pd.read_csv('tags.csv')

In [72]:
movies_join_ratings = ratings.join(movies.set_index('movieId'), on='movieId')
# movies_join_ratings

In [73]:
sum_rating_by_movies = movies_join_ratings.groupby(['movieId'])[['rating']].sum().sort_values('movieId')
sum_rating_by_movies = sum_rating_by_movies.reset_index()
# sum_rating_by_movies

In [74]:
count_rating_by_movies = movies_join_ratings.groupby(['movieId'])[['rating']].count().sort_values('movieId')
count_rating_by_movies = count_rating_by_movies.reset_index()
# count_rating_by_movies

In [75]:
avg_rating_by_movies = movies_join_ratings.groupby(['movieId'])[['rating']].mean().sort_values('movieId')
avg_rating_by_movies = avg_rating_by_movies.reset_index()
# avg_rating_by_movies

In [76]:
# добавляем rating, count, mean
moviess['sum'] = sum_rating_by_movies['rating']
moviess['count'] = count_rating_by_movies['rating']
moviess['mean'] = avg_rating_by_movies['rating']

moviess

movieId                                              title  \
0           1                                   Toy Story (1995)   
1           2                                     Jumanji (1995)   
2           3                            Grumpier Old Men (1995)   
3           4                           Waiting to Exhale (1995)   
4           5                 Father of the Bride Part II (1995)   
5           6                                        Heat (1995)   
6           7                                     Sabrina (1995)   
7           8                                Tom and Huck (1995)   
8           9                                Sudden Death (1995)   
9          10                                   GoldenEye (1995)   
10         11                     American President, The (1995)   
11         12                 Dracula: Dead and Loving It (1995)   
12         13                                       Balto (1995)   
13         14                                       Nixon (1995)   
14         15                            Cutthroat Island (1995)   
15         16                                      Casino (1995)   
16         17                       Sense and Sensibility (1995)   
17         18                                  Four Rooms (1995)   
18         19              Ace Ventura: When Nature Calls (1995)   
19         20                                 Money Train (1995)   
20         21                                  Get Shorty (1995)   
21         22                                     Copycat (1995)   
22         23                                   Assassins (1995)   
23         24                                      Powder (1995)   
24         25                           Leaving Las Vegas (1995)   
25         26                                     Othello (1995)   
26         27                                Now and Then (1995)   
27         28                                  Persuasion (1995)   
28         29  City of Lost Children, The (Cité des enfants p...   
29         30  Shanghai Triad (Yao a yao yao dao waipo qiao) ...   
...       ...                                                ...   
9712   188189                         Sorry to Bother You (2018)   
9713   188301                        Ant-Man and the Wasp (2018)   
9714   188675                                      Dogman (2018)   
9715   188751                Mamma Mia: Here We Go Again! (2018)   
9716   188797                                         Tag (2018)   
9717   188833              The Man Who Killed Don Quixote (2018)   
9718   189043                                  Boundaries (2018)   
9719   189111                                      Spiral (2018)   
9720   189333               Mission: Impossible - Fallout (2018)   
9721   189381                                    SuperFly (2018)   
9722   189547                                Iron Soldier (2010)   
9723   189713                              BlacKkKlansman (2018)   
9724   190183                           The Darkest Minds (2018)   
9725   190207                                        Tilt (2011)   
9726   190209                 Jeff Ross Roasts the Border (2017)   
9727   190213                                   John From (2015)   
9728   190215                                Liquid Truth (2017)   
9729   190219                                       Bunny (1998)   
9730   190221  Hommage à Zgougou (et salut à Sabine Mamou) (2...   
9731   191005                                     Gintama (2017)   
9732   193565                          Gintama: The Movie (2010)   
9733   193567  anohana: The Flower We Saw That Day - The Movi...   
9734   193571                                Silver Spoon (2014)   
9735   193573            Love Live! The School Idol Movie (2015)   
9736   193579           Jon Stewart Has Left the Building (2015)   
9737   193581          Black Butler: Book of the Atlantic (2017)   
9738   193583                       No Game No Life: Zero (201

In [77]:
moviess_join_ratings = ratings.join(moviess.set_index('movieId'), on='movieId')
# moviess_join_ratings

In [78]:
# выбираем фильмы для пользователя 1
moviess_users1 = moviess_join_ratings[moviess_join_ratings['userId']==1]
moviess_users1

userId  movieId  rating  timestamp  \
0         1        1     4.0  964982703   
1         1        3     4.0  964981247   
2         1        6     4.0  964982224   
3         1       47     5.0  964983815   
4         1       50     5.0  964982931   
5         1       70     3.0  964982400   
6         1      101     5.0  964980868   
7         1      110     4.0  964982176   
8         1      151     5.0  964984041   
9         1      157     5.0  964984100   
10        1      163     5.0  964983650   
11        1      216     5.0  964981208   
12        1      223     3.0  964980985   
13        1      231     5.0  964981179   
14        1      235     4.0  964980908   
15        1      260     5.0  964981680   
16        1      296     3.0  964982967   
17        1      316     3.0  964982310   
18        1      333     5.0  964981179   
19        1      349     4.0  964982563   
20        1      356     4.0  964980962   
21        1      362     5.0  964982588   
22        1      367     4.0  964981710   
23        1      423     3.0  964982363   
24        1      441     4.0  964980868   
25        1      457     5.0  964981909   
26        1      480     4.0  964982346   
27        1      500     3.0  964981208   
28        1      527     5.0  964984002   
29        1      543     4.0  964981179   
..      ...      ...     ...        ...   
202       1     3062     4.0  964981872   
203       1     3147     5.0  964983873   
204       1     3168     4.0  964982495   
205       1     3176     1.0  964983504   
206       1     3243     3.0  964981093   
207       1     3247     3.0  964983108   
208       1     3253     5.0  964981107   
209       1     3273     5.0  964983536   
210       1     3386     5.0  964983484   
211       1     3439     4.0  964981799   
212       1     3440     4.0  964981799   
213       1     3441     5.0  964982328   
214       1     3448     5.0  964984054   
215       1     3450     5.0  964981144   
216       1     3479     4.0  964981725   
217       1     3489     4.0  964981775   
218       1     3527     4.0  964982271   
219       1     3578     5.0  964980668   
220       1     3617     4.0  964980683   
221       1     3639     4.0  964982271   
222       1     3671     5.0  964981589   
223       1     3702     5.0  964983737   
224       1     3703     5.0  964981909   
225       1     3729     5.0  964982363   
226       1     3740     4.0  964982417   
227       1     3744     4.0  964980694   
228       1     3793     5.0  964981855   
229       1     3809     4.0  964981220   
230       1     4006     4.0  964982903   
231       1     5060     5.0  964984002   

                                                 title  \
0                                     Toy Story (1995)   
1                              Grumpier Old Men (1995)   
2                                          Heat (1995)   
3                          Seven (a.k.a. Se7en) (1995)   
4                           Usual Suspects, The (1995)   
5                           From Dusk Till Dawn (1996)   
6                                 Bottle Rocket (1996)   
7                                    Braveheart (1995)   
8                                       Rob Roy (1995)   
9                                Canadian Bacon (1995)   
10                                    Desperado (1995)   
11                                Billy Madison (1995)   
12                                       Clerks (1994)   
13              Dumb & Dumber (Dumb and Dumber) (1994)   
14                                      Ed Wood (1994)   
15           Star Wars: Episode IV - A New Hope (1977)   
16                                 Pulp Fiction (1994)   
17                                     Stargate (1994)   
18                                    Tommy Boy (1995)   
19                     Clear and Present Danger (1994)   
20                                 Forrest Gump (1994)   
21                             Jungle Book, The (1994)   
22  

In [79]:
del moviess_users1['userId']
del moviess_users1['timestamp']

In [80]:
moviess_users1

movieId  rating                                              title  \
0          1     4.0                                   Toy Story (1995)   
1          3     4.0                            Grumpier Old Men (1995)   
2          6     4.0                                        Heat (1995)   
3         47     5.0                        Seven (a.k.a. Se7en) (1995)   
4         50     5.0                         Usual Suspects, The (1995)   
5         70     3.0                         From Dusk Till Dawn (1996)   
6        101     5.0                               Bottle Rocket (1996)   
7        110     4.0                                  Braveheart (1995)   
8        151     5.0                                     Rob Roy (1995)   
9        157     5.0                              Canadian Bacon (1995)   
10       163     5.0                                   Desperado (1995)   
11       216     5.0                               Billy Madison (1995)   
12       223     3.0                                      Clerks (1994)   
13       231     5.0             Dumb & Dumber (Dumb and Dumber) (1994)   
14       235     4.0                                     Ed Wood (1994)   
15       260     5.0          Star Wars: Episode IV - A New Hope (1977)   
16       296     3.0                                Pulp Fiction (1994)   
17       316     3.0                                    Stargate (1994)   
18       333     5.0                                   Tommy Boy (1995)   
19       349     4.0                    Clear and Present Danger (1994)   
20       356     4.0                                Forrest Gump (1994)   
21       362     5.0                            Jungle Book, The (1994)   
22       367     4.0                                   Mask, The (1994)   
23       423     3.0                                  Blown Away (1994)   
24       441     4.0                          Dazed and Confused (1993)   
25       457     5.0                               Fugitive, The (1993)   
26       480     4.0                               Jurassic Park (1993)   
27       500     3.0                              Mrs. Doubtfire (1993)   
28       527     5.0                            Schindler's List (1993)   
29       543     4.0                So I Married an Axe Murderer (1993)   
..       ...     ...                                                ...   
202     3062     4.0                            Longest Day, The (1962)   
203     3147     5.0                             Green Mile, The (1999)   
204     3168     4.0                                  Easy Rider (1969)   
205     3176     1.0                    Talented Mr. Ripley, The (1999)   
206     3243     3.0                                  Encino Man (1992)   
207     3247     3.0                                  Sister Act (1992)   
208     3253     5.0                               Wayne's World (1992)   
209     3273     5.0                                    Scream 3 (2000)   
210     3386     5.0                                         JFK (1991)   
211     3439     4.0  Teenage Mutant Ninja Turtles II: The Secret of...   
212     3440     4.0            Teenage Mutant Ninja Turtles III (1993)   
213     3441     5.0                                    Red Dawn (1984)   
214     3448     5.0                       Good Morning, Vietnam (1987)   
215     3450     5.0                              Grumpy Old Men (1993)   
216     3479     4.0                                   Ladyhawke (1985)   
217     3489     4.0                                        Hook (1991)   
218     3527     4.0                                    Predator (1987)   
219     3578     5.0                                   Gladiator (2000)   
220     3617     4.0                                   Road Trip (2000)   
221     3639     4.0                Man with the Golden Gun, The (1974)   
222     3671     5.0                             Blazing Saddles (1974)   
223     3702     5.0          

In [81]:
moviess_users1_join_tags = moviess_users1.join(tags.set_index('movieId'), on='movieId')

In [82]:
# moviess_users1_join_tags
moviess_users1_join_tags.tag.unique().shape

(414,)

In [83]:
tag_strings = []

for movie, group in moviess_users1_join_tags.sort_values('movieId').groupby('movieId'):
    tag_strings.append(' '.join([str(s).replace(' ', '').replace('-', '') for s in group.tag.values]))

In [84]:
tag_strings

['pixar pixar fun',
 'moldy old',
 'nan',
 'mystery twistending serialkiller',
 'heist twistending tricky thriller suspense mindfuck',
 'nan',
 'crime offbeatcomedy quirky',
 'Scotland revenge Oscar(BestCinematography) melgibson swordfight inspirational historical epic beautifulscenery Medieval',
 'nan',
 'nan',
 'nan',
 'school AdamSandler stoplookingatmeswan',
 'generationX quirky witty hilarious cynical independentfilm',
 'nan',
 'moviebusiness',
 'StarWars ROBOTSANDANDROIDS space classicscifi scifi StarWars darthvader lukeskywalker spaceopera scifi spaceadventure oldiebutgoodie spaceepic action classicscifi Nerd classic EPIC engrossingadventure classicscifi greatsoundtrack EPIC scifi spaceaction classic spaceaction',
 "nonlinear nonlineartimeline nonlinear philosophical neonoir multiplestorylines nonlinearnarrative noir nonlinearstoryline originalplot notablesoundtrack offensive organisedcrime organizedcrime original multiplestories outoforder Palmed'Or parody nonlineartimeline mot

In [85]:
tfidf_vectorizer_tags = TfidfVectorizer()
X_train_tfidf_tags = tfidf_vectorizer_tags.fit_transform(tag_strings)

In [86]:
# векторы тегов
tfidfs_on_tag = X_train_tfidf_tags.toarray()
tfidfs_on_tag

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [87]:
tfidfs_on_tag.shape[1]

395

даже если брать теги только к фильмам первого юзера получилось 395, а это всё равно больше,чем фильмов первого юзера (их 232)